pkg/helpers/status.go

```diff
@@ -0,0 +1,228 @@
+		helper.Info("workflow node is in failed state")
+		err = helper.Remediating(ctx, instance)
+		return true, time.Duration(0), err
+	case v1alpha12.NodeSucceeded:
+		if _, err := helper.Succeeded(ctx, instance); err != nil {
+			return false, time.Duration(0), err
+		}
+		helper.V(1).Info("workflow has succeeded")
+		requeueTime = v1alpha1.GetInterval(instance) // update the requeue time if we succeeded
+	default:
+		if instance.GetReadyCondition() == meta.RollingBackReason {
+			if err := helper.RollingBack(ctx, instance); err != nil {
+				return false, time.Duration(0), err
+			}
+		}
+		helper.V(1).Info("workflow is still progressing")
+	}
+	return false, requeueTime, nil
+}
+
+func (helper *StatusHelper) UpdateStatus(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	chartConditionMap, subChartConditionMap, err := helper.marshallChartStatus(ctx, instance)
+	if err != nil {
+		return err
+	}
+	instance.Status.Applications = getAppStatus(instance, chartConditionMap, subChartConditionMap)
+	instance.DeploySucceeded()
+
+	return helper.patchStatus(ctx, instance)
+}
+
+func (helper *StatusHelper) Failed(ctx context.Context, instance *v1alpha1.ApplicationGroup, err error) (ctrl.Result, error) {
+	helper.Recorder.Event(instance, "Warning", "ReconcileError", fmt.Sprintf("Failed to reconcile ApplicationGroup %v with Error : %v", instance.Name, err))
+	instance.ReadyFailed(err.Error())
+	instance.DeployFailed(err.Error())
+
+	return ctrl.Result{}, helper.patchStatus(ctx, instance)
+}
+
+func (helper *StatusHelper) Succeeded(ctx context.Context, instance *v1alpha1.ApplicationGroup) (ctrl.Result, error) {
+	// Set the status conditions into a succeeding state
+	instance.ReadySucceeded()
+	instance.DeploySucceeded()
+	if err := helper.patchStatus(ctx, instance); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Set the last successful annotation for rollback scenarios
+	instance.SetLastSuccessful()
+	if err := helper.Patch(ctx, instance, helper.PatchFrom); err != nil {
+		helper.V(1).Error(err, "failed to patch the application group annotations")
+		return ctrl.Result{}, err
+	}
+	helper.Recorder.Event(instance, "Normal", "ReconcileSuccess", fmt.Sprintf("Successfully reconciled ApplicationGroup %s", instance.Name))
+	return ctrl.Result{}, nil
+}
+
+func (helper *StatusHelper) Remediating(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	instance.ReadyFailed("workflow in failed state, starting rollback...")
+	instance.DeployFailed("workflow in failed state, starting rollback...")
+
+	return helper.patchStatus(ctx, instance)
+}
+
+// RollingBack sets the meta.ReadyCondition to 'True' and
+// meta.RollingBack reason and message
+func (helper *StatusHelper) RollingBack(ctx context.Context, instance *v1alpha1.ApplicationGroup) error {
+	meta.SetResourceCondition(instance, meta.ReadyCondition, metav1.ConditionTrue, meta.FailedReason, "workflow failed because of helmreleases during upgrade, rolling back...")
```

Comment:
Is it always because of HelmRelease failures (as the error message states) or can it also be a workflow failure?